schema/function.go

```diff
@@ -0,0 +1,326 @@
+// inputs. The return type of the handler should be any.
+// - The handler function handles execution of the function.
+// - The typeHandler function returns the output type given the input type. If the inputs are invalid, the
+// handler should return an error.
+func NewDynamicCallableFunction(
+	id string,
+	inputs []Type,
+	display Display,
+	handler any,
+	typeHandler func(inputType []Type) (Type, error),
+) (CallableFunction, error) {
+	parsedHandler := reflect.ValueOf(handler)
+	// Validate the input types match the provided ones.
+	err := validateInputTypeCompatibility(inputs, parsedHandler)
+	if err != nil {
+		return nil, err
+	}
+	// Validate the output type
+	returnCount := parsedHandler.Type().NumOut()
+
+	switch {
+	case returnCount != 2:
+		return nil, fmt.Errorf("expected dynamic handler to have two returns, one any and one error, but got %d return types", returnCount)
+	case parsedHandler.Type().Out(1).Name() != errorType:
+		return nil, fmt.Errorf("expected additional return type to be an error return, but got %s", parsedHandler.Type().Out(1).Name())
+	case parsedHandler.Type().Out(0).Kind() != reflect.Interface:
+		return nil, fmt.Errorf("expected 'any' return type for handler, but got %s", parsedHandler.Type().Out(0))
+	}
+	return &CallableFunctionSchema{
+		IDValue:            id,
+		InputsValue:        inputs,
+		DefaultOutputValue: nil,
+		DisplayValue:       display,
+		Handler:            parsedHandler,
+		DynamicTypeHandler: typeHandler,
+	}, nil
+}
+
+func validateInputTypeCompatibility(
+	inputs []Type,
+	handler reflect.Value,
+) error {
+	// Validate the input types match the provided ones.
+	specifiedParams := len(inputs)
+	actualParams := handler.Type().NumIn()
+	if specifiedParams != actualParams {
+		return fmt.Errorf(
+			"parameter inputs do not match handler inputs. handler has %d, expected %d",
+			actualParams, specifiedParams)
+	}
+	for i := 0; i < len(inputs); i++ {
+		expectedType := inputs[i].ReflectedType()
+		handlerType := handler.Type().In(i)
+		if expectedType != handlerType {
+			return fmt.Errorf(
+				"type mismatch for parameter at index %d. handler has %v, inputs specifies %v",
+				i, handlerType, expectedType)
+		}
+	}
+	return nil
+}
+
+type FunctionSchema struct {
+	IDValue      string  `json:"id"`
+	InputsValue  []Type  `json:"inputs"`
+	OutputValue  Type    `json:"output"`
+	DisplayValue Display `json:"display"`
+}
+
+func (f FunctionSchema) ID() string {
+	return f.IDValue
+}
+
+func (f FunctionSchema) Parameters() []Type {
+	return f.InputsValue
+}
+
+func (f FunctionSchema) Output(_ []Type) (Type, error) {
+	return f.OutputValue, nil
+}
+
+func (f FunctionSchema) Display() Display {
+	return f.DisplayValue
+}
+
+func (f FunctionSchema) String() string {
+	result := f.ID() + "("
+	for i := 0; i < len(f.Parameters()); i++ {
+		if i != 0 {
+			result += ", "
+		}
+		result += string(f.Parameters()[i].TypeID())
+	}
+	result += ") "
```

Comment:
> it would be more work to make that work

Yes it would, the _first_ time...but after that, you reap the rewards from the investment.

That is, if you had a `Map()` function, then I believe that you could code this like this:
```go
	guts := Map(f.Parameters(), func(p Type) string { return string(p.TypeID()) })
	result := f.ID() + "(" + ", ".join(guts) + ") "
```
which is probably better code (shorter, fewer moving parts, DRYer, better encapsulation, yadda yadda yadda).

The `Map()` function is easy to write (there are examples [here](https://stackoverflow.com/questions/33726731/short-way-to-apply-a-function-to-all-elements-in-a-list-in-golang), including a pointer to a package that you could import), and once it is available to your code, you might find other applications of it (even beyond `CallableFunctionSchema`).